src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+{
+    Task<RequestContextType?> CreateRequestContextAsync(IQueueItem<RequestContextType> queueItem, CancellationToken cancellationToken);
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
+
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(
+        bool mutatesSolutionState,
+        bool requiresLSPSolution,
+        IRequestHandler<TRequestType, TResponseType, RequestContextType> handler,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
+        string methodName,
+        CancellationToken cancellationToken);
+
+    void Shutdown();
+}
+
+public interface IQueueItem<RequestContextType> where RequestContextType : struct
+{
+    /// <summary>
+    /// Begins executing the work specified by this queue item.
+    /// </summary>
+    Task CallbackAsync(RequestContextType? context, CancellationToken cancellationToken);
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.RequiresLSPSolution" />
+    bool RequiresLSPSolution { get; }
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.MutatesSolutionState" />
+    bool MutatesSolutionState { get; }
+
+    string MethodName { get; }
+
+    /// <summary>
+    /// The document identifier that will be used to find the solution and document for this request. This comes from the <see cref="TextDocumentIdentifier"/> returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType, TResponseContextType}.GetTextDocumentIdentifier(RequestType)"/>.
+    /// </summary>
+    TextDocumentIdentifier? TextDocument { get; }
+
+    /// <inheritdoc cref="RequestContext.ClientCapabilities" />
+    ClientCapabilities ClientCapabilities { get; }
+
+    /// <summary>
+    /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
+    /// tracing/logging subsystem.
+    /// </summary>
+    Guid ActivityId { get; }
+
+    IRequestMetrics Metrics { get; }
+}
+
+public interface IRequestMetrics
+{
+    void RecordCancellation();
+    void RecordExecutionStart();
+    void RecordFailure();
+    void RecordSuccess();
+}
+
+public record RequestHandlerMetadata(string MethodName, Type RequestType, Type ResponseType);
+
+public interface IRequestDispatcher<RequestContextType> where RequestContextType : struct
+{
+    ImmutableArray<RequestHandlerMetadata> GetRegisteredMethods();
+
+    Task<TResponseType?> ExecuteRequestAsync<TRequestType, TResponseType>(
+        string methodName,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
+        IRequestExecutionQueue<RequestContextType> queue,
+        CancellationToken cancellationToken);
+}
+
+public abstract class LanguageServerTarget<RequestContextType> : ILanguageServer where RequestContextType : struct
+{
+    private readonly ICapabilitiesProvider _capabilitiesProvider;
+
+    protected readonly JsonRpc _jsonRpc;
```

Comment:
Is this protected so that consumers can do server -> client communication?

Reply:
I forget offhand why it's protected. In a lot of these places I just got sloppy with access modifiers while investigating dead ends.

Reply:
If possible (totally might not be possible), it'd be awesome to hide the JsonRpc impl from handlers and have them call into a specific set of services that are only allowed to access this object. in the LSP platform they let everyone party on their jsonRpc objects and it was a huuuge problem

Reply:
Looking at this again the next day y'all are of course both correct, any spike that involves making JsonRpc non-private is a bad idea to start with lol.

I think the specific problem I was trying to solve was getting the jsonRpc `IClientLanguageServerManager` in Roslyn, but rather than polluting everything with that I should just construct the things that need it in the Roslyn specific constructor.